python-checks/src/main/java/org/sonar/python/checks/UndeclaredNameUsageCheck.java

```diff
@@ -74,21 +76,27 @@ public void initialize(Context context) {
       }
       DefinedVariablesAnalysis analysis = DefinedVariablesAnalysis.analyze(cfg, functionDef.localVariables());
       Set<CfgBlock> unreachableBlocks = CfgUtils.unreachableBlocks(cfg);
-      cfg.blocks().forEach(block -> checkCfgBlock(block, ctx, analysis.getDefinedVariables(block), unreachableBlocks, analysis));
+      cfg.blocks().forEach(block -> checkCfgBlock(block, ctx, analysis.getDefinedVariables(block), unreachableBlocks, analysis, ignoredSymbols));
     });
   }
 
-  private static void checkCfgBlock(
-    CfgBlock cfgBlock, SubscriptionContext ctx, DefinedVariables definedVariables, Set<CfgBlock> unreachableBlocks, DefinedVariablesAnalysis analysis) {
+  private static void checkCfgBlock(CfgBlock cfgBlock, SubscriptionContext ctx, DefinedVariables definedVariables,
+                                    Set<CfgBlock> unreachableBlocks, DefinedVariablesAnalysis analysis, List<Symbol> ignoredSymbols) {
     Map<Symbol, DefinedVariablesAnalysis.VariableDefinition> currentState = new HashMap<>(definedVariables.getIn());
     for (Tree element : cfgBlock.elements()) {
       definedVariables.getVariableUsages(element).forEach((symbol, symbolUsage) -> {
         if (symbolUsage.isWrite()) {
           currentState.put(symbol, DefinedVariablesAnalysis.VariableDefinition.DEFINED);
         }
         DefinedVariablesAnalysis.VariableDefinition varDef = currentState.getOrDefault(symbol, DefinedVariablesAnalysis.VariableDefinition.DEFINED);
-        if (symbolUsage.isRead() && isUndefined(varDef) && !isSymbolUsedInUnreachableBlocks(analysis, unreachableBlocks, symbol) && !isParameter(element)) {
-          ctx.addIssue(element, symbol.name() + " is used before it is defined. Move the definition before.");
+        if (symbolUsage.isRead() && isUndefined(varDef) && !isSymbolUsedInUnreachableBlocks(analysis, unreachableBlocks, symbol) && !isParameter(element)
+          && !ignoredSymbols.contains(symbol)) {
+          ignoredSymbols.add(symbol);
+          Optional<Usage> suspectUsage = symbol.usages().stream().filter(u -> TreeUtils.hasDescendant(element, t -> t.equals(u.tree()))).findFirst();
```

Comment:
This filtering is quite suspicious to me. @andrea-guarino-sonarsource can you please have  a look?

Reply:
From what I understand, such filtering is required because the `UsageVisitor` associates a `Tree` element with a map of its `Symbol` and `SymbolUsage`. However, we are missing the precise link with the actual `Usage` within the `Tree` (as it is most likely a descendant of the tree rather than the tree itself).

I feels to me that in order to avoid this kind of filtering, I would therefore need to refactor the `UsageVisitor`, which is maybe a bit overkill if done just for this case (but maybe it would be helpful in other cases too?).

Reply:
I agree it's not easy to understand why we need this filtering. As discussed, I'd suggest adding a property on `SymbolUsage` to get the corresponding usages.